include/rmm/mr/memory_resource.hpp

```diff
@@ -0,0 +1,427 @@
+                             std::size_t alignment = alignof(std::max_align_t)) = 0;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Compare this resource to another.
+   *
+   * Two host_memory_resources compare equal if and only if memory allocated
+   * from one host_memory_resource can be deallocated from the other and vice
+   * versa.
+   *
+   * By default, simply checks if \p *this and \p other refer to the same
+   * object, i.e., does not check if they are two objects of the same class.
+   *
+   * @param other The other resource to compare to
+   * @return true If the two resources are equivalent
+   * @return false If the two resources are not equal
+   *---------------------------------------------------------------------------**/
+  virtual bool do_is_equal(memory_resource const& other) const noexcept
+  {
+    return this == &other;
+  }
+};
+
+/**
+ * @brief Base class for all multi-stream RMM memory resources.
+ *
+ * @tparam kind   The kind of memory
+ *
+ * This is based on `std::pmr::memory_resource`:
+ * https://en.cppreference.com/w/cpp/memory/memory_resource
+ *
+ * When C++17 is available for use in RMM, `rmm::host_memory_resource` should
+ * inherit from `std::pmr::memory_resource`.
+ *
+ * This class serves as the interface for all memory resources that do not need
+ * to work with multiple CUDA streams.
+ *
+ * There are two private, pure virtual functions that all derived classes must
+ * implement: `do_allocate` and `do_deallocate`. Optionally, derived classes may
+ * also override `is_equal`. By default, `is_equal` simply performs an identity
+ * comparison.
+ *
+ * The public, non-virtual functions `allocate`, `deallocate`, and `is_equal`
+ * simply call the private virtual functions. The reason for this is to allow
+ * implementing shared, default behavior in the base class. For example, the
+ * base class' `allocate` function may log every allocation, no matter what
+ * derived class implementation is used.
+ */
+template <memory_kind kind>
+class stream_aware_memory_resource : public memory_resource<kind> {
+ public:
+  using memory_resource<kind>::allocate;
+  using memory_resource<kind>::deallocate;
+
+  /**
+   * @brief Allocates memory of size at least \p bytes.
+   *
+   * The returned pointer is aligned at the defautl alignment for the memory kind
+   * allocated by this resource.
+   *
+   * The memory returned must be available for immediate use on given stream.
+   * If an implementation does not support streams, it should return memory which
+   * can be used on any stream.
+   *
+   * @throws `rmm::bad_alloc` When the requested `bytes` cannot be allocated on
+   * the specified `stream`.
+   *
+   * @param bytes The size of the allocation
+   * @param stream Stream on which to perform allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* allocate(std::size_t bytes, cuda_stream_view stream)
```

Comment:
I think since we are going with `do_allocate_async` we should mirror it in the public interface as well. That would make it more explicit as well. 
```suggestion
  void* allocate_async(std::size_t bytes, cuda_stream_view stream)
```

Reply:
Done.